pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
+		serviceHostname := serviceHostname(serviceName, serviceNamespace)
+
+		if _, found := certificateChain.CertificatesIssued[serviceHostname]; !found {
+			certificateChain.CertificatesIssued[serviceHostname] = newCertificateIssue(serviceName, serviceNamespace)
 		}
 
-		for _, clientConfig := range m.clientConfigList(webhook) {
-			// Update the CA bundle at webhook
-			updatedCABundle, err := updateCABundle(clientConfig.CABundle)
-			if err != nil {
-				return errors.Wrap(err, "failed updating CA bundle")
-			}
-			clientConfig.CABundle = updatedCABundle
+		caBundleName := caBundleName(object.key.String(), name)
+		certificateChain.CertificatesIssued[serviceHostname].CACertPEM[caBundleName] = config.CABundle
+		key := newObjectKey("Secret", serviceNamespace, serviceName)
+		if _, found := objects[key]; !found {
+			objects[key] = &keyedObject{key, nil}
 		}
+	}
+}
 
-		err = m.client.Update(context.TODO(), webhook)
-		if err != nil {
-			return err
+// mapWebhookToChain maps a webhook object from certificate chain data.
+func mapWebhookFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	clientConfigList := clientConfigMap(object.kobject)
+	for name, config := range clientConfigList {
+		serviceHostname := serviceHostname(config.Service.Name, config.Service.Namespace)
+		certificateIssue := certificateChain.CertificatesIssued[serviceHostname]
+		if certificateIssue == nil {
+			continue
 		}
-		return nil
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to update webhook CABundle")
+		caBundleName := caBundleName(object.key.String(), name)
+		caBundle := certificateIssue.CACertPEM[caBundleName]
+		if caBundle == nil {
+			continue
+		}
+		config.CABundle = caBundle
+	}
+}
+
+// mapWebhookToChain maps a secret object to certificate chain data.
+func mapSecretToChain(object *keyedObject, objects objectMap, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretToChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretToChain(object, certificateChain)
+}
+
+// mapWebhookToChain maps a secret object from certificate chain data.
+func mapSecretFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretFromChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretFromChain(object, certificateChain)
+}
+
+func mapServiceSecretToChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	key := secret.Data[corev1.TLSPrivateKeyKey]
+	cert := secret.Data[corev1.TLSCertKey]
+	if key == nil || cert == nil {
+		return
+	}
+
+	name := serviceHostname(object.key.NamespacedName.Name, object.key.NamespacedName.Namespace)
+
+	certificateChain.CertificatesIssued[name].KeyPEM = key
+	certificateChain.CertificatesIssued[name].CertPEM = cert
```

Comment:
I suppose we don't want to do the decoding into x506.Certificate becase we want to decouple between PEM (which is an standard way of represent this info) and the golang structs, but I am not sure that we need this decoupling.

If we do the decoding here the chain constructors will be massively simplified.